File: resonagraph/core/phase_key.py
"""
Phase key management for cryptographic access control.

According to copilot-instructions.md Section 5:
- Root K_root derives per-topic K_topic via HKDF
- Rotate keys every 24h for security
- Without correct phase key K, probes misalign (R < R_min)
"""

import hashlib
import hmac
import secrets
from typing import Optional


class PhaseKey:
    """
    Manages cryptographic phase keys for access control.
    
    According to design.md Section 6.1:
    - Enforcement: Without K, probes misalign
    - Key Hierarchy: K_root derives K_topic via HKDF
    - Rotate every 24h
    """
    
    KEY_SIZE = 32  # 256 bits
    
    def __init__(self, key_bytes: bytes):
        """
        Initialize a phase key from raw bytes.
        
        Args:
            key_bytes: Raw key material (should be 32 bytes)
        """
        if len(key_bytes) != self.KEY_SIZE:
            raise ValueError(f"Key must be exactly {self.KEY_SIZE} bytes")
        self._key = key_bytes
    
    @classmethod
    def from_secret(cls, secret: bytes) -> "PhaseKey":
        """
        Create a phase key from a secret.
        
        Args:
            secret: Secret bytes (will be hashed to proper size)
            
        Returns:
            PhaseKey instance
        """
        # Hash secret to get key material
        key_bytes = hashlib.sha256(secret).digest()
        return cls(key_bytes)
    
    @classmethod
    def generate(cls) -> "PhaseKey":
        """
        Generate a new random phase key.
        
        Returns:
            PhaseKey instance with cryptographically secure random bytes
        """
        key_bytes = secrets.token_bytes(cls.KEY_SIZE)
        return cls(key_bytes)
    
    @classmethod
    def derive_topic_key(
        cls,
        root_key: "PhaseKey",
        topic: str,
        salt: Optional[bytes] = None
    ) -> "PhaseKey":
        """
        Derive a per-topic key from root key using HKDF.
        
        According to design.md Section 6.1:
        K_root derives per-topic K_topic via HKDF
        
        Args:
            root_key: Root phase key
            topic: Topic string for key derivation
            salt: Optional salt (random if not provided)
            
        Returns:
            Derived topic key
        """
        if salt is None:
            salt = secrets.token_bytes(16)
        
        # HKDF Extract: PRK = HMAC-Hash(salt, IKM)
        prk = hmac.new(salt, root_key._key, hashlib.sha256).digest()
        
        # HKDF Expand: OKM = HMAC-Hash(PRK, info || 0x01)
        info = topic.encode('utf-8')
        okm = hmac.new(prk, info + b'\x01', hashlib.sha256).digest()
        
        return cls(okm)
    
    def get_bytes(self) -> bytes:
        """
        Get the raw key bytes.
        
        Returns:
            Raw key material
        """
        return self._key
    
    def __repr__(self) -> str:
        """String representation (does not reveal key)."""
        return f"PhaseKey(***hidden***)"
    
    def __eq__(self, other: object) -> bool:
        """
        Compare keys in constant time.
        
        According to copilot-instructions.md:
        Use constant-time comparisons for security-critical paths
        """
        if not isinstance(other, PhaseKey):
            return False
        
        # Constant-time comparison using hmac.compare_digest
        return hmac.compare_digest(self._key, other._key)
    
    def __hash__(self) -> int:
        """Hash for use in sets/dicts."""
        return hash(self._key)

File: javascript/src/core/PhaseKey.ts
/**
 * Phase Key module for cryptographic access control.
 * 
 * According to design.md Section 5.1:
 * - HMAC-based cryptographic binding
 * - Key hierarchy via HKDF
 * - Rotate keys every 24h for security
 */

import * as crypto from 'crypto';

// Constants
const DEFAULT_KEY_LENGTH = 32; // 256 bits
const HKDF_ALGORITHM = 'sha256';
const HMAC_ALGORITHM = 'sha256';

export class PhaseKey {
  private keyBytes: Buffer;
  private createdAt: Date;

  /**
   * Create a PhaseKey from raw bytes.
   * 
   * @param keyBytes - 32-byte secret key
   */
  constructor(keyBytes: Buffer) {
    if (keyBytes.length !== DEFAULT_KEY_LENGTH) {
      throw new Error(`PhaseKey must be ${DEFAULT_KEY_LENGTH} bytes, got ${keyBytes.length}`);
    }
    this.keyBytes = keyBytes;
    this.createdAt = new Date();
  }

  /**
   * Create a PhaseKey from a secret buffer.
   * 
   * @param secret - Secret buffer (will be hashed if not 32 bytes)
   * @returns New PhaseKey instance
   */
  static fromSecret(secret: Buffer): PhaseKey {
    if (secret.length === DEFAULT_KEY_LENGTH) {
      return new PhaseKey(secret);
    }
    
    // Hash the secret to get 32 bytes
    const hash = crypto.createHash('sha256');
    hash.update(secret);
    return new PhaseKey(hash.digest());
  }

  /**
   * Create a PhaseKey from a passphrase string.
   * 
   * @param passphrase - String passphrase
   * @returns New PhaseKey instance
   */
  static fromPassphrase(passphrase: string): PhaseKey {
    return PhaseKey.fromSecret(Buffer.from(passphrase, 'utf-8'));
  }

  /**
   * Generate a random PhaseKey.
   * 
   * @returns New random PhaseKey instance
   */
  static generate(): PhaseKey {
    return new PhaseKey(crypto.randomBytes(DEFAULT_KEY_LENGTH));
  }

  /**
   * Derive a child key using HKDF.
   * 
   * @param info - Context information for key derivation
   * @returns New derived PhaseKey
   */
  deriveKey(info: string): PhaseKey {
    const salt = Buffer.alloc(DEFAULT_KEY_LENGTH); // Zero salt for deterministic derivation
    const derivedKey = crypto.hkdfSync(
      HKDF_ALGORITHM,
      this.keyBytes,
      salt,
      Buffer.from(info, 'utf-8'),
      DEFAULT_KEY_LENGTH
    );
    return new PhaseKey(Buffer.from(derivedKey));
  }

  /**
   * Compute HMAC for phase encoding.
   * 
   * @param data - Data to authenticate
   * @returns HMAC digest
   */
  computeHMAC(data: Buffer): Buffer {
    const hmac = crypto.createHmac(HMAC_ALGORITHM, this.keyBytes);
    hmac.update(data);
    return hmac.digest();
  }

  /**
   * Get the raw key bytes.
   * 
   * @returns Key bytes (copy)
   */
  getBytes(): Buffer {
    return Buffer.from(this.keyBytes);
  }

  /**
   * Get key creation timestamp.
   * 
   * @returns Creation date
   */
  getCreatedAt(): Date {
    return this.createdAt;
  }

  /**
   * Check if key should be rotated (older than 24 hours).
   * 
   * @returns true if key should be rotated
   */
  shouldRotate(): boolean {
    const age = Date.now() - this.createdAt.getTime();
    const DAY_MS = 24 * 60 * 60 * 1000;
    return age > DAY_MS;
  }
}
